File: concore.java
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.channels.FileChannel;
import java.nio.channels.FileLock;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.List;

/**
 * Java implementation of concore local communication.
 *
 * This class provides file-based inter-process communication for control systems,
 * mirroring the functionality of concore.py.
 */
public class concore {
    private static final ConcoreJavaRuntimeCore runtime = new ConcoreJavaRuntimeCore("./in", "./out", true);

    private static final Path BASE_DIR = Paths.get("").toAbsolutePath().normalize();
    private static final Path PID_REGISTRY_FILE = BASE_DIR.resolve("concorekill_pids.txt");
    private static final Path KILL_SCRIPT_FILE = BASE_DIR.resolve("concorekill.bat");

    // initialize on class load, same as Python module-level init
    static {
        if (isWindows()) {
            registerPid();
            writeKillScript();
            Runtime.getRuntime().addShutdownHook(new Thread(concore::cleanupPid));
        }
        Runtime.getRuntime().addShutdownHook(new Thread(concore::terminateZmq));
    }

    private static boolean isWindows() {
        String os = System.getProperty("os.name");
        return os != null && os.toLowerCase().contains("win");
    }

    private static void registerPid() {
        try {
            String pid = String.valueOf(ProcessHandle.current().pid());
            try (FileChannel channel = FileChannel.open(PID_REGISTRY_FILE,
                    StandardOpenOption.CREATE, StandardOpenOption.WRITE, StandardOpenOption.APPEND)) {
                try (FileLock lock = channel.lock()) {
                    channel.write(ByteBuffer.wrap((pid + System.lineSeparator()).getBytes(java.nio.charset.StandardCharsets.UTF_8)));
                }
            }
        } catch (IOException e) {
        }
    }

    private static void cleanupPid() {
        String pid = String.valueOf(ProcessHandle.current().pid());
        if (!Files.exists(PID_REGISTRY_FILE)) return;
        List<String> remaining = new ArrayList<>();
        try (FileChannel channel = FileChannel.open(PID_REGISTRY_FILE,
                StandardOpenOption.READ, StandardOpenOption.WRITE)) {
            try (FileLock lock = channel.lock()) {
                ByteBuffer buf = ByteBuffer.allocate((int) channel.size());
                channel.read(buf);
                buf.flip();
                String content = java.nio.charset.StandardCharsets.UTF_8.decode(buf).toString();
                for (String line : content.split("\\R")) {
                    String trimmed = line.trim();
                    if (!trimmed.isEmpty() && !trimmed.equals(pid)) {
                        remaining.add(trimmed);
                    }
                }
                channel.truncate(0);
                channel.position(0);
                if (!remaining.isEmpty()) {
                    StringBuilder sb = new StringBuilder();
                    for (String r : remaining) sb.append(r).append(System.lineSeparator());
                    channel.write(ByteBuffer.wrap(sb.toString().getBytes(java.nio.charset.StandardCharsets.UTF_8)));
                }
            }
        } catch (IOException e) {
        }
        if (remaining.isEmpty()) {
            try { Files.deleteIfExists(PID_REGISTRY_FILE); } catch (IOException e) {}
            try { Files.deleteIfExists(KILL_SCRIPT_FILE); } catch (IOException e) {}
        }
    }

    private static void writeKillScript() {
        try {
            String regName = PID_REGISTRY_FILE.getFileName().toString();
            String batName = KILL_SCRIPT_FILE.getFileName().toString();
            String script = "@echo off\r\n";
            script += "if not exist \"%~dp0" + regName + "\" (\r\n";
            script += "    echo No PID registry found. Nothing to kill.\r\n";
            script += "    exit /b 0\r\n";
            script += ")\r\n";
            script += "for /f \"usebackq tokens=*\" %%p in (\"%~dp0" + regName + "\") do (\r\n";
            script += "    wmic process where \"ProcessId=%%p\" get CommandLine /value 2>nul | find /i \"concore\" >nul\r\n";
            script += "    if not errorlevel 1 (\r\n";
            script += "        echo Killing concore process %%p\r\n";
            script += "        taskkill /F /PID %%p >nul 2>&1\r\n";
            script += "    ) else (\r\n";
            script += "        echo Skipping PID %%p - not a concore process or not running\r\n";
            script += "    )\r\n";
            script += ")\r\n";
            script += "del /q \"%~dp0" + regName + "\" 2>nul\r\n";
            script += "del /q \"%~dp0" + batName + "\" 2>nul\r\n";
            Files.write(KILL_SCRIPT_FILE, script.getBytes(java.nio.charset.StandardCharsets.UTF_8));
        } catch (IOException e) {
        }
    }

    /**
     * Sets maxtime from concore.maxtime file, or uses defaultValue if file not found.
     * Catches both IOException and RuntimeException to match Python safe_literal_eval.
     */
    public static void defaultMaxTime(double defaultValue) {
        runtime.defaultMaxTime(defaultValue);
    }

    // package-level helpers for testing with temp directories
    static void setInPath(String path) { runtime.setInPath(path); }
    static void setOutPath(String path) { runtime.setOutPath(path); }
    static void setDelay(int ms) { runtime.setDelay(ms); }
    static double getSimtime() { return runtime.getSimtime(); }
    static void resetState() { runtime.resetState(); }

    public static boolean unchanged() {
        return runtime.unchanged();
    }

    public static Object tryParam(String n, Object i) {
        return runtime.tryParam(n, i);
    }

    /**
     * Reads data from a port file. Returns the values after extracting simtime.
     * Input format: [simtime, val1, val2, ...]
     * Returns: list of values after simtime
     * Includes max retry limit to avoid infinite blocking (matches Python behavior).
     */
    public static ReadResult read(int port, String name, String initstr) {
        return convertResult(runtime.readFilePort(port, name, initstr));
    }

    /**
     * Writes data to a port file.
     * Prepends simtime+delta to the value list, then serializes to Python-literal format.
     * Accepts List or String values (matching Python implementation).
     */
    public static void write(int port, String name, Object val, int delta) {
        runtime.writeFilePort(port, name, val, delta);
    }

    /**
     * Parses an initial value string like "[0.0, 1.0, 2.0]".
     * Extracts simtime from position 0 and returns the remaining values as a List.
     */
    public static List<Object> initVal(String simtimeVal) {
        return runtime.initVal(simtimeVal);
    }

    public static void initZmqPort(String portName, String portType, String address, String socketTypeStr) {
        runtime.initZmqPort(portName, portType, address, socketTypeStr);
    }

    public static void terminateZmq() {
        runtime.terminateZmq();
    }

    /**
     * Reads data from a ZMQ port. Same wire format as file-based read:
     * expects [simtime, val1, val2, ...], strips simtime, returns the rest.
     */
    public static ReadResult read(String portName, String name, String initstr) {
        return convertResult(runtime.readZmqPort(portName, name, initstr));
    }

    /**
     * Writes data to a ZMQ port. Prepends [simtime+delta] to match file-based write behavior.
     */
    public static void write(String portName, String name, Object val, int delta) {
        runtime.writeZmqPort(portName, name, val, delta);
    }

    /**
     * Parses a Python-literal string into Java objects using a recursive descent parser.
     * Supports: dict, list, int, float, string (single/double quoted), bool, None, nested structures.
     * This replaces the broken split-based parser that could not handle quoted commas or nesting.
     */
    static Object literalEval(String s) {
        return ConcoreJavaRuntimeCore.literalEval(s);
    }

    private static ReadStatus convertStatus(ConcoreJavaRuntimeCore.ReadStatus status) {
        return ReadStatus.valueOf(status.name());
    }

    private static ReadResult convertResult(ConcoreJavaRuntimeCore.ReadResult result) {
        return new ReadResult(convertStatus(result.status), result.data);
    }

    public enum ReadStatus {
        SUCCESS, FILE_NOT_FOUND, TIMEOUT, PARSE_ERROR, RETRIES_EXCEEDED
    }

    public static class ReadResult {
        public final ReadStatus status;
        public final List<Object> data;
        ReadResult(ReadStatus status, List<Object> data) {
            this.status = status;
            this.data = data;
        }
    }

}

File: concore.py
import time
import logging
import os
import atexit
from ast import literal_eval
import sys
import re
import zmq
import numpy as np
import signal

import concore_base

logger = logging.getLogger('concore')
logger.addHandler(logging.NullHandler())

#these lines mute the noisy library
logging.getLogger('matplotlib').setLevel(logging.WARNING)
logging.getLogger('PIL').setLevel(logging.WARNING)
logging.getLogger('urllib3').setLevel(logging.WARNING) 
logging.getLogger('requests').setLevel(logging.WARNING) 


# if windows, register this process PID for safe termination
# Previous approach: single "concorekill.bat" overwritten by each node (race condition).
# New approach: append PID to shared registry; generate validated kill script.
# See: https://github.com/ControlCore-Project/concore/issues/391

_PID_REGISTRY_FILE = "concorekill_pids.txt"
_KILL_SCRIPT_FILE = "concorekill.bat"

def _register_pid():
    """Append current PID to the shared registry file."""
    try:
        with open(_PID_REGISTRY_FILE, "a") as f:
            f.write(str(os.getpid()) + "\n")
    except OSError:
        pass

def _cleanup_pid():
    """Remove current PID from registry on exit. Uses file locking on Windows."""
    pid = str(os.getpid())
    try:
        if not os.path.exists(_PID_REGISTRY_FILE):
            return
        with open(_PID_REGISTRY_FILE, "r+") as f:
            if hasattr(sys, 'getwindowsversion'):
                import msvcrt
                msvcrt.locking(f.fileno(), msvcrt.LK_LOCK, 1)
            pids = [line.strip() for line in f if line.strip()]
            remaining = [p for p in pids if p != pid]
            if remaining:
                f.seek(0)
                f.truncate()
                for p in remaining:
                    f.write(p + "\n")
            else:
                f.close()
                try:
                    os.remove(_PID_REGISTRY_FILE)
                except OSError:
                    pass
                try:
                    os.remove(_KILL_SCRIPT_FILE)
                except OSError:
                    pass
    except OSError:
        pass

def _write_kill_script():
    """Generate concorekill.bat that validates each PID before killing."""
    try:
        script = "@echo off\r\n"
        script += 'if not exist "%~dp0' + _PID_REGISTRY_FILE + '" (\r\n'
        script += "    echo No PID registry found. Nothing to kill.\r\n"
        script += "    exit /b 0\r\n"
        script += ")\r\n"
        script += 'for /f "usebackq tokens=*" %%p in ("%~dp0' + _PID_REGISTRY_FILE + '") do (\r\n'
        script += '    tasklist /FI "PID eq %%p" 2>nul | find /i "python" >nul\r\n'
        script += "    if not errorlevel 1 (\r\n"
        script += "        echo Killing Python process %%p\r\n"
        script += "        taskkill /F /PID %%p >nul 2>&1\r\n"
        script += "    ) else (\r\n"
        script += "        echo Skipping PID %%p - not a Python process or not running\r\n"
        script += "    )\r\n"
        script += ")\r\n"
        script += 'del /q "%~dp0' + _PID_REGISTRY_FILE + '" 2>nul\r\n'
        script += 'del /q "%~dp0' + _KILL_SCRIPT_FILE + '" 2>nul\r\n'
        with open(_KILL_SCRIPT_FILE, "w", newline="") as f:
            f.write(script)
    except OSError:
        pass

if hasattr(sys, 'getwindowsversion'):
    _register_pid()
    _write_kill_script()
    atexit.register(_cleanup_pid)

ZeroMQPort = concore_base.ZeroMQPort
convert_numpy_to_python = concore_base.convert_numpy_to_python
safe_literal_eval = concore_base.safe_literal_eval
parse_params = concore_base.parse_params

# Global variables
zmq_ports = {}
_cleanup_in_progress = False

last_read_status = "SUCCESS"

s = ''
olds = ''
delay = 1
retrycount = 0
inpath = "./in" #must be rel path for local
outpath = "./out"
simtime = 0

def _port_path(base, port_num):
    return base + str(port_num)

concore_params_file = os.path.join(_port_path(inpath, 1), "concore.params")
concore_maxtime_file = os.path.join(_port_path(inpath, 1), "concore.maxtime")

# Load input/output ports if present
iport = safe_literal_eval("concore.iport", {})
oport = safe_literal_eval("concore.oport", {})

_mod = sys.modules[__name__]

# ===================================================================
# ZeroMQ Communication Wrapper
# ===================================================================
def init_zmq_port(port_name, port_type, address, socket_type_str):
    concore_base.init_zmq_port(_mod, port_name, port_type, address, socket_type_str)

def terminate_zmq():
    """Clean up all ZMQ sockets and contexts before exit."""
    concore_base.terminate_zmq(_mod)

def signal_handler(sig, frame):
    """Handle interrupt signals gracefully."""
    print(f"\nReceived signal {sig}, shutting down gracefully...")
    try:
        atexit.unregister(terminate_zmq)
    except Exception:
        pass
    concore_base.terminate_zmq(_mod)
    sys.exit(0)

# Register cleanup handlers
atexit.register(terminate_zmq)
signal.signal(signal.SIGINT, signal_handler)   # Handle Ctrl+C
if not hasattr(sys, 'getwindowsversion'):
    signal.signal(signal.SIGTERM, signal_handler)  # Handle termination (Unix only)

params = concore_base.load_params(concore_params_file)

#9/30/22
def tryparam(n, i):
    """Return parameter `n` from params dict, else default `i`."""
    return params.get(n, i)

#9/12/21
# ===================================================================
# Simulation Time Handling
# ===================================================================
def default_maxtime(default):
    """Read maximum simulation time from file or use default."""
    global maxtime
    maxtime = safe_literal_eval(concore_maxtime_file, default)

default_maxtime(100)

def unchanged():
    """Check if global string `s` is unchanged since last call."""
    return concore_base.unchanged(_mod)

# ===================================================================
# I/O Handling (File + ZMQ)
# ===================================================================
def read(port_identifier, name, initstr_val):
    """Read data from a ZMQ port or file-based port.

    Returns:
        tuple: (data, success_flag) where success_flag is True if real
            data was received, False if a fallback/default was used.
            Also sets ``concore.last_read_status`` to one of:
            SUCCESS, FILE_NOT_FOUND, TIMEOUT, PARSE_ERROR,
            EMPTY_DATA, RETRIES_EXCEEDED.

    Backward compatibility:
        Legacy callers that do ``value = concore.read(...)`` will
        receive a tuple.  They can adapt with::

            result = concore.read(...)
            if isinstance(result, tuple):
                value, ok = result
            else:
                value, ok = result, True

        Alternatively, check ``concore.last_read_status`` after the
        call.
    """
    global last_read_status
    result = concore_base.read(_mod, port_identifier, name, initstr_val)
    last_read_status = concore_base.last_read_status
    return result


def write(port_identifier, name, val, delta=0):
    concore_base.write(_mod, port_identifier, name, val, delta)

def initval(simtime_val_str): 
    return concore_base.initval(_mod, simtime_val_str)
